### ISSUE TITLE
Pods from ReplicationController are sometimes not removed

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#30990]
I'm currently debugging failures from scalability suite.

It basically started failing on this run:
http://kubekins.dls.corp.google.com:8080/view/Scalability/job/kubernetes-e2e-gce-scalability/10148/

This suite is running WITHOUT garbage collector.

There are tens of failures like this one:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/load.go:197 deleting rc load-small-rc-46 Expected error: <*errors.errorString | 0xc8275da7c0>: { s: "error while waiting for pods to become inactive load-small-rc-46: timed out waiting for the condition", } error while waiting for pods to become inactive load-small-rc-46: timed out waiting for the condition not to have occurred /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/load.go:345
```

I'm also able to reproduce this problem from head on my local 100-node cluster.

We did some debugging of it and it seems that because of some reason, pods from some replication controllers are simply not being removed.
What happens is that:
1. we call: DeleteRCAndPods https://github.com/kubernetes/kubernetes/blob/master/test/e2e/load.go#L345
2. This should result in:
- scaling RC to 0
- waiting until all pods are gone
- removing replication controller

However, what happens (looking into logs) is that the first two doesn't happen, but replication controller is indeed removed. From RC logs (from some run):
```
I0819 11:12:44.621239       7 replication_controller.go:585] Finished syncing controller "e2e-tests-load-30-nodepods-1-ikyk3/load-small-rc-226" (5.195731ms)
I0819 11:17:30.168494       7 replication_controller.go:598] Replication Controller has been deleted e2e-tests-load-30-nodepods-1-ikyk3/load-small-rc-226
I0819 11:17:30.168508       7 replication_controller.go:585] Finished syncing controller "e2e-tests-load-30-nodepods-1-ikyk3/load-small-rc-226" (16.194�s)
```

The timing when it started happening matches exactly merge of #30670, but I don't know how this can affect it.

@caesarxuchao @lavalamp @gmarek 

http://kubekins.dls.corp.google.com:8080/view/Scalability/job/kubernetes-e2e-gce-scalability/10148/ is the good run to start debugging with

**Comment by caesarxuchao** [comment]
I think this bug is less critical. The test doesn't enable GC. #30670 just changed some test verification in e2e tests. The worst we have to do it to revert part of #30670. I'll look into this later.

**Comment by wojtek-t** [comment]
@caesarxuchao - I don't think it's less critical. I'm seeing that pods are not being removed even though they were in the past. It's not a verification issue - when I reproduced it locally it just didn't work. It's real issue.

**Comment by caesarxuchao** [comment]
One hypothesis based on your debugging.

Before #30670, this test:
1. calls reaper, which in turn calls scaler to scale down the rc, 
2. scaler waits RC controller to update rc.status.replicas to 0, which means there is no active pods,
3. reaper deletes the rc
3. test calls `waitForRCPodsGone`, which is buggy, it always returns true immediately. The reason is `waitForRCPodsGone` calls the List on PodStore without waiting for the reflector to call List().

After #30670, this test:
1. same
2. same
3. same
4. calls `waitForPodInactive`, which should return immediately, because in step 2 should guarantee there is no active pods. But the test timed out here. So it seems there is problem with the reaper or the rc controller.

**Comment by wojtek-t** [comment]
Also from our discussion offline: "currently we think that it's not a new bug, it was just revelead by @caesarxuchao PR, and it has always been there".

**Comment by krousey** [comment]
Just as an FYI, 1.4 release coordinator is citing the scale test failures as a blocker for cutting an alpha release.

cc: @pwittrock

**Comment by bprashanth** [comment]
my unconfirmed guess is that the client (reaper/scale) is deleting RC for some generation that it hasn't fulfilled. See discussion on:  https://github.com/kubernetes/kubernetes/issues/25170, https://github.com/kubernetes/kubernetes/pull/25171. If that turns out the be the issue, we need to change how generation is exported from controller which is probably not going to happen this release.

**Comment by caesarxuchao** [comment]
Discussed offline, the prematurely updated observedGeneration shouldn't be a problem here. Because apart from the observerGeneration, the reaper also waits for the rc.Status.Relicas==0. Some logs show the RC has at least 2 active replicas before the deletion happen, so the prematurely updated observedGeneration is not enough to trigger the flake.

**Comment by bprashanth** [comment]
> Some logs show the RC has at least 2 active replicas before the deletion happen, 

This is not sufficient, you need to confirm that the rc wasn't scaled to 0. We could always have logged 2 replicas, scaled the rc to 0 then up to 5, and you get status.replicas=0 generation=(generation for 5). It sounds like we should start with debuggin when it regressed and boil it down to a smaller test case than load.go.

**Comment by caesarxuchao** [comment]
Forget to mention the test scales the particular RC from 5 to 2, then to 3, then delete. I saw the log message generated by the rc manager saying "Too few "e2e-tests-load-30-nodepods-1-tokqg"/"load-small-rc-269" replicas, need 3, creating 1". I think that guarantees the status.replicas is not 0 at the time of deletion?

**Comment by bprashanth** [comment]
Check out the scale loop: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/scale.go#L191

And note that:
* you need status = spec
* you need a higher generation
* the rc is from a watcher 

Rapid scale up/down is going to confuse it, assuming that's the scaler you're talking about in previous comments.

**Comment by caesarxuchao** [comment]
I probably found the cause:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/scale.go#L180

The scale watch is started with ListOptions.ResourceVersion=0, which means the watch is started from a random resource version, so the scale loop might just get a watch event from the history where all conditions were met.

**Comment by bprashanth** [comment]
Just size down to 0 in your test and wait out the rc, if that's what you really want. The scaler has a harder time because it doesn't know the target size, so it ends up doing things like this. If you're going to scale rapidly, do what deployment does, scale to X, wait for X pods, take action={scale again, delete etc}.

**Comment by smarterclayton** [comment]
I suspect it's 0 because of "too old" errors.  Which once scale has been executed is unlikely to fail, so we may want to guard against that.

**Comment by caesarxuchao** [comment]
@smarterclayton sorry, could you explain more? I don't follow.

I'm going to take the RV from the response of the RC update and use it in the Watch.

**Comment by smarterclayton** [comment]
At one point someone probably started watching before scale was sent, so the resource version for the RC was <<< current index resulting in an etcd "watch index too old".  So the fix was to watch from zero (or we never had the RC).  Your suggested fix I think is correct.

**Comment by wojtek-t** [comment]
I agree that the fix is correct, however, I'm not 100% convinced that this is a real issue.
In all failed test runs that I've seen, the RC for which deletion we were waiting had 0 replicas only on startup which happened minutes before we were deleting it (e.g. 5 minutes before). That said, I really don't believe that "List" was returning element from 5 minutes ago.

That said - I agree that what you're running is correct, but I'm afraid it won't fix the failures...

**Comment by caesarxuchao** [comment]
We need to monitor the 100-node test to see if this is actually fixed by #31051.
